libqtile/widget/statusnotifier.py

```diff
@@ -0,0 +1,897 @@
+        self.bus = await MessageBus().connect()
+        for iface in BUS_NAMES:
+            w = StatusNotifierWatcher(iface)
+            w.on_item_added = self.add_item
+            w.on_item_removed = self.remove_item
+            await w.start()
+
+            # Not quite following spec here as we're not registering
+            # the host on the bus.
+            w.RegisterStatusNotifierHost(self.name)
+            self.watchers.append(w)
+
+    def item_added(self, item, service, future):
+        success = future.result()
+        # If StatusNotifierItem object was created successfully then we
+        # add to our list and redraw the bar
+        if success:
+            self.items.append(item)
+            self.widget.bar.draw()
+
+    def add_item(self, service, path=None):
+        """
+        Creates a StatusNotifierItem for the given service and tries to
+        start it.
+        """
+        item = StatusNotifierItem(
+            self.bus, service,
+            path=path,
+            display_menu_callback=self.display_menu_callback,
+            icon_theme=self.icon_theme
+        )
+        item.on_icon_changed = self.widget.bar.draw
+        if item not in self.items:
+            task = asyncio.create_task(item.start())
+            task.add_done_callback(
+                partial(self.item_added, item, service)
+            )
+
+    def remove_item(self, interface):
+        # Check if the interface is in out list of items and, if so,
+        # remove it and redraw the bar
+        if interface in self.items:
+            self.items.remove(interface)
+            self.widget.bar.draw()
+
+
+host = StatusNotifierHost()  # noqa: E303
+
+
+class StatusNotifier(base._Widget):
+    """
+    A 'system tray' widget using the freedesktop StatusNotifierItem
+    specification.
+
+    Letf-clicking an icon will trigger an activate event.
+    Right-clicking an icon will display a context menu where provided
+    by the application.
+    """
+
+    orientations = base.ORIENTATION_BOTH
+
+    defaults = [
+        ('icon_size', 16, 'Icon width'),
+        ('icon_theme', None, 'Name of theme to use for app icons'),
+        ('padding', 5, 'Padding between icons'),
+        ('font', 'sans', 'Font for menu text')
+    ]
+
+    def __init__(self, **config):
+        base._Widget.__init__(self, bar.CALCULATED, **config)
+        self.add_defaults(StatusNotifier.defaults)
+        self.add_callbacks(
+            {
+                "Button1": self.activate,
+            }
+        )
+
+    @property
+    def available_icons(self):
+        return [item for item in self.host.items if item.has_icons]
+
+    def calculate_length(self):
+        if not self.host.items:
+            return 0
+
+        return (
+            len(self.available_icons) * (self.icon_size + self.padding)
+            + self.padding
+        )
+
+    def _configure(self, qtile, bar):
+        if self.configured:
+            return
+
+        self.host = host
```

Comment:
This seems redundant as `host` can be accessed everywhere already. Is this necessary?

Reply:
I think we both know the answer to that is 'no'!